syntax = "proto3";

package bench.proto;

message TraceLine {
  message Malloc {
    optional uint64 result_id = 1;
    optional uint64 input_size = 2;
    // Optional, if given, this is an aligned alloc.
    optional uint64 input_alignment = 3;
  }

  message Calloc {
    optional uint64 result_id = 1;
    optional uint64 input_nmemb = 2;
    optional uint64 input_size = 3;
  }

  message Realloc {
    optional uint64 result_id = 1;
    optional uint64 input_id = 2;
    optional uint64 input_size = 3;
  }

  message Free {
    optional uint64 input_id = 1;
    optional uint64 input_size_hint = 2;
    optional uint64 input_alignment_hint = 3;
  }

  oneof op {
    Malloc malloc = 1;
    Calloc calloc = 4;
    Realloc realloc = 3;
    Free free = 2;
  }
}

message Tracefile {
  repeated TraceLine lines = 1;

  // The maximum number of simultaneously allocated items in this trace.
  optional uint64 max_simultaneous_allocs = 2;
}
